bookkeeper-server/src/main/java/org/apache/bookkeeper/sasl/SaslServerState.java

```diff
@@ -0,0 +1,294 @@
+
+            if (configurationEntries == null) {
+                String errorMessage = "Could not find a '" + configurationEntry + "' entry in this configuration: Server cannot start.";
+
+                throw new IOException(errorMessage);
+            }
+            credentials.clear();
+            for (AppConfigurationEntry entry : configurationEntries) {
+                Map<String, ?> options = entry.getOptions();
+                // Populate DIGEST-MD5 user -> password map with JAAS configuration entries from the "Server" section.
+                // Usernames are distinguished from other options by prefixing the username with a "user_" prefix.
+                for (Map.Entry<String, ?> pair : options.entrySet()) {
+                    String key = pair.getKey();
+                    if (key.startsWith(USER_PREFIX)) {
+                        String userName = key.substring(USER_PREFIX.length());
+                        credentials.put(userName, (String) pair.getValue());
+                    }
+                }
+            }
+        }
+
+        public void handle(Callback[] callbacks) throws UnsupportedCallbackException {
+            for (Callback callback : callbacks) {
+                if (callback instanceof NameCallback) {
+                    handleNameCallback((NameCallback) callback);
+                } else if (callback instanceof PasswordCallback) {
+                    handlePasswordCallback((PasswordCallback) callback);
+                } else if (callback instanceof RealmCallback) {
+                    handleRealmCallback((RealmCallback) callback);
+                } else if (callback instanceof AuthorizeCallback) {
+                    handleAuthorizeCallback((AuthorizeCallback) callback);
+                }
+            }
+        }
+
+        private void handleNameCallback(NameCallback nc) {
+            // check to see if this user is in the user password database.
+            if (credentials.get(nc.getDefaultName()) == null) {
+                LOG.error("User '" + nc.getDefaultName() + "' not found in list of JAAS DIGEST-MD5 users.");
+                return;
+            }
+            nc.setName(nc.getDefaultName());
+            userName = nc.getDefaultName();
+        }
+
+        private void handlePasswordCallback(PasswordCallback pc) {
+            if (credentials.containsKey(userName)) {
+                pc.setPassword(credentials.get(userName).toCharArray());
+            } else {
+                LOG.info("No password found for user: " + userName);
+            }
+        }
+
+        private void handleRealmCallback(RealmCallback rc) {
+            LOG.debug("client supplied realm: " + rc.getDefaultText());
+            rc.setText(rc.getDefaultText());
+        }
+
+        private void handleAuthorizeCallback(AuthorizeCallback ac) {
+            String authenticationID = ac.getAuthenticationID();
+            String authorizationID = ac.getAuthorizationID();
```

Comment:
Thank you @EronWright 
This patch has already been integrated into main codebase.
I think that "impersionation" will be very interesting once we introduce ACLs or quotas....
Can you file an issue in order to track better your idea ?